patches/server/0950-Paper-Item-API.patch

```diff
@@ -0,0 +1,4333 @@
++            case PINK -> Material.PINK_BANNER;
++            case GRAY -> Material.GRAY_BANNER;
++            case LIGHT_GRAY -> Material.LIGHT_GRAY_BANNER;
++            case CYAN -> Material.CYAN_BANNER;
++            case PURPLE -> Material.PURPLE_BANNER;
++            case BLUE -> Material.BLUE_BANNER;
++            case BROWN -> Material.BROWN_BANNER;
++            case GREEN -> Material.GREEN_BANNER;
++            case RED -> Material.RED_BANNER;
++            case BLACK -> Material.BLACK_BANNER;
++        };
++    }
++
++    @Override
++    public @NotNull MetaBlockState clone() {
++        return (MetaBlockState) super.clone();
++    }
++
++    @Override
++    void migrateFrom(PropertyHolder.Mutable mutable) {
++        this.propertyHolder.unset(ItemProperties.BLOCK_ENTITY_TAG);
++    }
++}
+diff --git a/src/main/java/io/papermc/paper/item/meta/MetaBook.java b/src/main/java/io/papermc/paper/item/meta/MetaBook.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..1a9478b599fbaad898f05aafbcaba8610c2ce07a
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/item/meta/MetaBook.java
+@@ -0,0 +1,471 @@
++package io.papermc.paper.item.meta;
++
++import com.google.common.collect.ImmutableList;
++import io.papermc.paper.inventory.item.properties.ItemProperties;
++import io.papermc.paper.property.MutablePropertyHolderImpl;
++import io.papermc.paper.property.PropertyHolder;
++import io.papermc.paper.property.PropertyKey;
++import io.papermc.paper.property.value.ListValue;
++import io.papermc.paper.property.value.Value;
++import net.kyori.adventure.text.Component;
++import net.kyori.adventure.text.serializer.legacy.LegacyComponentSerializer;
++import net.md_5.bungee.api.chat.BaseComponent;
++import net.md_5.bungee.chat.ComponentSerializer;
++import org.apache.commons.lang.Validate;
++import org.bukkit.FireworkEffect;
++import org.bukkit.craftbukkit.inventory.CraftMetaBook;
++import org.bukkit.craftbukkit.util.CraftChatMessage;
++import org.bukkit.inventory.meta.BookMeta;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++import org.spigotmc.ValidateUtils;
++
++import java.util.AbstractList;
++import java.util.Arrays;
++import java.util.List;
++
++// TODO: Cleanup, I hate all of this
++public class MetaBook extends PaperMeta implements BookMeta {
++
++    static final int MAX_PAGES = 100;
++    static final int MAX_PAGE_LENGTH = 320; // 256 limit + 64 characters to allow for psuedo colour codes
++    static final int MAX_TITLE_LENGTH = 32;
++
++    PropertyKey<ListValue<Component>> PAGES_WRITTEN_STRING_WRAPPER = PropertyKey.ofList("pages", Component.class);
++
++    public MetaBook(PropertyHolder.Mutable propertyHolder) {
++        super(propertyHolder);
++    }
++
++    @Override
++    public boolean hasAuthor() {
++        return MetaBook.this.propertyHolder.hasProperty(ItemProperties.AUTHOR);
++    }
++
++    @Override
++    public boolean hasTitle() {
```

Comment:
Most of the simpler methods could be moved into API as we can operate on.

```java
ItemStack#getPropertyHolder()..hasProperty(ItemProperties.AUTHOR);
```

Then we're able to fully unit test that on the API side as we just inject a dummy PropertyHolder.

Reply:
That's true, although that will most likely be far future.